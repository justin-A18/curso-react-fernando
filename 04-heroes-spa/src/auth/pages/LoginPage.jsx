import { useContext } from 'react';
import { useNavigate } from 'react-router-dom';
import { AuthContext } from '../context';

export const LoginPage = () => {
	const navigate = useNavigate();

	const { login } = useContext(AuthContext);

	const handleLogin = () => {
		const lastPath = localStorage.getItem('lastPath') || '/';

		login('Justin');

		navigate(lastPath, {
			replace: true,
		});
	};

	return (
		<div className='container mt-5'>
			<h1>Login</h1>
			<hr />

			<button
				className='btn btn-primary'
				onClick={handleLogin}>
				Login
			</button>
		</div>
	);
};
