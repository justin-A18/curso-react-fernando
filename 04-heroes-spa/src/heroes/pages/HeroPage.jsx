import { Navigate, useNavigate, useParams } from 'react-router-dom';
import { getHeroById } from '../helpers/getHeroById';
import { useMemo } from 'react';

export const HeroPage = () => {
	// ObtÃ©n el id de la URL usando useParams
	// Nos sirve para obtener los parametros

	const { id } = useParams();
	//Memorizar valores, Cada que cambie le id
	const hero = useMemo(() => getHeroById(id), [id]);
	const navigate = useNavigate();

	const handleNavigateReturn = () => {
		//Navega al historial anterior
		navigate(-1);
	};

	if (!hero) return <Navigate to='/marvel' />;

	return (
		<div className='row mt-5'>
			<div className='col-4 animate__animated animate__fadeInLeft'>
				<img
					src={`/heroes/${id}.jpg`}
					alt={hero.superhero}
					className='img-thumbnail'
				/>
			</div>

			<div className='col-8 animate__animated animate__fadeInRight'>
				<h3>{hero.superhero}</h3>
				<ul className='list-group list-group-flush'>
					<li className='list-group-item'>
						Alter Ego: <b>{hero.alter_ego}</b>
					</li>

					<li className='list-group-item'>
						Publisher: <b>{hero.publisher}</b>
					</li>

					<li className='list-group-item'>
						Firts Appearance: <b>{hero.first_appearance}</b>
					</li>
				</ul>

				<h5 className='mt-3'>Characters: {hero.characters}</h5>

				<button
					className='btn btn-primary'
					onClick={handleNavigateReturn}>
					Regresar
				</button>
			</div>
		</div>
	);
};
