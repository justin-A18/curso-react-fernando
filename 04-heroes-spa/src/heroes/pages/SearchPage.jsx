import { useLocation, useNavigate } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import queryString from 'query-string';
import { getHeroByName } from '../helpers/getHeroByName';
import { HeroCard } from '../components/HeroCard';

export const SearchPage = () => {
	const navigate = useNavigate(); // obtener navegacion
	const location = useLocation(); // obtener la localizacion donde nos encontramos

	//Libreria para simplificar la extraccion de los querys: npm install query-string
	const { q = '' } = queryString.parse(location.search);
	const heroes = getHeroByName(q);

	const showSearch = q.length === 0;
	const showError = q.length > 0 && heroes.length === 0;

	const { searchText, onInputChange } = useForm({
		searchText: q,
	});

	const handleFormSubmit = (e) => {
		e.preventDefault();
		if (searchText.trim().length <= 1) return;

		navigate(`?q=${searchText}`);
	};

	return (
		<>
			<h1>Search</h1>
			<hr />

			<div className='row'>
				<div className='col-5'>
					<h4>Searching</h4>
					<hr />

					<form onSubmit={handleFormSubmit} aria-label='form-search'>
						<input
							type='text'
							placeholder='Search a hero'
							className='form-control'
							name='searchText'
							value={searchText}
							onChange={onInputChange}
							autoComplete='off'
						/>

						<button className='btn btn-outline-primary mt-1'>Search</button>
					</form>
				</div>

				<div className='col-7'>
					<h4>Results</h4>
					<hr />

					<div
						className='alert alert-primary'
						style={{ display: showSearch ? '' : 'none' }}
						aria-label='search-hero'>
						Search a hero
					</div>

					<div
						className='alert alert-danger'
						aria-label='alert-danger'
						style={{ display: showError ? '' : 'none' }}>
						No hero with <b>{q}</b>
					</div>

					{heroes.map((heroe) => (
						<HeroCard
							key={heroe.id}
							{...heroe}
						/>
					))}

					{/*<*/}
				</div>
			</div>
		</>
	);
};
