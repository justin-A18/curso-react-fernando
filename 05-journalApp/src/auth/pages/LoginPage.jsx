/* eslint-disable no-extra-boolean-cast */
import { useDispatch, useSelector } from 'react-redux';
import { Link as RouterLink } from 'react-router-dom';

import { Google } from '@mui/icons-material';
import {
	Alert,
	Button,
	Grid,
	Link,
	TextField,
	Typography,
} from '@mui/material';

import { AuthLayout } from '../layout/AuthLayout';
import { useForm } from '../../hooks/useForm';
import {
	startGoogleSignIn,
	startLoginWithEmailPassword,
} from '../../store/auth';
import { useMemo } from 'react';

const formData = {
	email: '',
	password: '',
};

export const LoginPage = () => {
	const { email, password, onInputChange } = useForm(formData);
	const { status, errorMessage } = useSelector((state) => state.auth);
	const dispatch = useDispatch();

	const isAuthenticated = useMemo(() => status === 'checking', [status]);

	const handleSubmit = (e) => {
		e.preventDefault();

		dispatch(startLoginWithEmailPassword({ email, password }));
	};

	const onGoogleSignIn = () => {
		dispatch(startGoogleSignIn());
	};

	return (
		<AuthLayout title='Login'>
			<form
				aria-label='submit-form'
				onSubmit={handleSubmit}
				className='animate__animated animate__fadeIn animate__faster'>
				<Grid container>
					<Grid
						item
						xs={12}
						sx={{ mt: 2 }}>
						<TextField
							label='correo'
							type='email'
							name='email'
							placeholder='correo@gmail.com'
							inputProps={{
								'data-testid': 'email'
							}}
							fullWidth
							onChange={onInputChange}
						/>
					</Grid>

					<Grid
						item
						xs={12}
						sx={{ mt: 2 }}>
						<TextField
							label='contraseÃ±a'
							type='password'
							inputProps={{
								'data-testid': 'password'
							}}
							name='password'
							placeholder='password'
							fullWidth
							onChange={onInputChange}
						/>
					</Grid>

					<Grid
						container
						display={!!errorMessage ? '' : 'none'}
						sx={{ mt: 1 }}>
						<Grid
							item
							xs={12}>
							<Alert severity='error'>{errorMessage}</Alert>
						</Grid>
					</Grid>

					<Grid
						container
						spacing={2}
						alignItems='center'
						sx={{ mt: 1 }}>
						<Grid
							item
							xs={12}
							md={6}>
							<Button
								disabled={isAuthenticated}
								type='submit'
								variant='contained'
								aria-label='login-btn'
								fullWidth>
								Login
							</Button>
						</Grid>

						<Grid
							item
							xs={12}
							md={6}>
							<Button
								disabled={isAuthenticated}
								variant='contained'
								fullWidth
								aria-label='google-btn'
								onClick={onGoogleSignIn}>
								<Google />
								<Typography sx={{ ml: 1 }}>Google</Typography>
							</Button>
						</Grid>
					</Grid>

					<Grid
						container
						justifyContent='end'
						direction='row'
						sx={{ mt: 2 }}>
						<Link
							component={RouterLink}
							color='inherit'
							to='/auth/register'>
							Crea una cuenta
						</Link>
					</Grid>
				</Grid>
			</form>
		</AuthLayout>
	);
};
