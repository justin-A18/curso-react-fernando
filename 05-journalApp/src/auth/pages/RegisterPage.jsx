/* eslint-disable no-extra-boolean-cast */
import {
	Alert,
	Button,
	Grid,
	Link,
	TextField,
	Typography,
} from '@mui/material';
import { Link as RouterLink } from 'react-router-dom';
import { AuthLayout } from '../layout/AuthLayout';
import { useForm } from '../../hooks';
import { useMemo, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { startCreatingUserWithEmailPassword } from '../../store/auth/thunks';

const formData = {
	name: '',
	email: '',
	password: '',
};

const formValidations = {
	name: {
		isValid: (value = '') => value.length >= 1,
		errorMessage: 'El nombre es obligatorio',
	},
	email: {
		isValid: (value = '') => value.includes('@'),
		errorMessage: 'El correo debe tener un arroba',
	},
	password: {
		isValid: (value = '') => value.length >= 6,
		errorMessage: 'la contrase単a debe tener minimo 6 caracteres',
	},
};

export const RegisterPage = () => {
	const [formSubmitted, setFormSubmitted] = useState(false);

	const dispatch = useDispatch();
	const { status, errorMessage } = useSelector((state) => state.auth);
	const isCheckingAuthentication = useMemo(
		() => status === 'checking',
		[status]
	);

	const {
		name,
		email,
		password,
		nameValid,
		emailValid,
		isFormValid,
		passwordValid,
		onResetForm,
		onInputChange,
	} = useForm(formData, formValidations);
	const handleSubmit = (e) => {
		e.preventDefault();

		if (!isFormValid) {
			setFormSubmitted(true);
			return;
		}

		setFormSubmitted(false);

		dispatch(
			startCreatingUserWithEmailPassword({ displayName: name, email, password })
		);

		onResetForm();
	};

	return (
		<AuthLayout title='Crear Cuenta'>
			<form
				onSubmit={handleSubmit}
				className='animate__animated animate__fadeIn animate__faster'>
				<Grid container>
					<Grid
						item
						xs={12}
						sx={{ mt: 2 }}>
						<TextField
							fullWidth
							type='text'
							name='name'
							label='Nombre Completo'
							placeholder='Justin Huertas'
							value={name}
							onChange={onInputChange}
							error={!!nameValid && formSubmitted}
							helperText={!!nameValid && formSubmitted ? nameValid : null}
						/>
					</Grid>

					<Grid
						item
						xs={12}
						sx={{ mt: 2 }}>
						<TextField
							fullWidth
							type='email'
							name='email'
							label='Correo'
							placeholder='correo@gmail.com'
							value={email}
							onChange={onInputChange}
							error={!!emailValid && formSubmitted}
							helperText={!!emailValid && formSubmitted ? emailValid : null}
						/>
					</Grid>

					<Grid
						item
						xs={12}
						sx={{ mt: 2 }}>
						<TextField
							fullWidth
							type='password'
							name='password'
							label='Contrase単a'
							placeholder='contrase単a'
							value={password}
							onChange={onInputChange}
							error={!!passwordValid && formSubmitted}
							helperText={
								!!passwordValid && formSubmitted ? passwordValid : null
							}
						/>
					</Grid>

					<Grid
						container
						spacing={2}
						sx={{ mt: 1 }}>
						<Grid
							item
							xs={12}
							display={!!errorMessage ? '' : 'none'}>
							<Alert severity='error'>{errorMessage}</Alert>
						</Grid>

						<Grid
							item
							xs={12}>
							<Button
								type='submit'
								disabled={isCheckingAuthentication}
								variant='contained'
								fullWidth>
								Crear Cuenta
							</Button>
						</Grid>
					</Grid>

					<Grid
						container
						direction='row'
						justifyContent='end'
						sx={{ mt: 2 }}>
						<Typography sx={{ mr: 1 }}>多Ya tienes una cuenta?</Typography>
						<Link
							component={RouterLink}
							color='inherit'
							to='/auth/login'>
							Ingresar
						</Link>
					</Grid>
				</Grid>
			</form>
		</AuthLayout>
	);
};
