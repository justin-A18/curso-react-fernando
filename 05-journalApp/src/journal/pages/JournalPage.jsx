import { IconButton } from '@mui/material';
import { JournalLayout } from '../layout/JournalLayout';
import { NoteView, NothinSelectedView } from '../views';
import { AddOutlined } from '@mui/icons-material';
import { useDispatch, useSelector } from 'react-redux';
import { startNewNote } from '../../store/journal/thunks';

export const JournalPage = () => {
	const dispatch = useDispatch();
	const { isSaving, active } = useSelector((state) => state.journal);

	const onclickNewNote = () => {
		dispatch(startNewNote());
	};

	return (
		<JournalLayout>
			{active ? <NoteView /> : <NothinSelectedView />}

			<IconButton
				onClick={onclickNewNote}
				size='large'
				disabled={isSaving}
				sx={{
					color: 'white',
					backgroundColor: 'error.main',
					':hover': { backgroundColor: 'error.main', opacity: 0.9 },
					':disabled': {
						backgroundColor: 'error.main',
						opacity: 0.9,
						color: 'white',
					},
					position: 'fixed',
					right: 50,
					bottom: 50,
				}}>
				<AddOutlined sx={{ fontSize: 30 }} />
			</IconButton>
		</JournalLayout>
	);
};
