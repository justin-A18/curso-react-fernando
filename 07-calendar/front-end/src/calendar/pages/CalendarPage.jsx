import { useEffect, useState } from 'react';

import { Calendar } from 'react-big-calendar';
import 'react-big-calendar/lib/css/react-big-calendar.css';

import { useUiActions } from '../../hooks/useUiActions';
import { localizer, getMessagesEs } from '../../helpers';
import { useCalendarActions } from '../../hooks/useCalendarActions';
import {
	NavBar,
	CalendarEvent,
	CalendarModal,
	FabAddNew,
	FabDelete,
} from '../';

const initialStateLastView = localStorage.getItem('lastView') || 'month';

export const CalendarPage = () => {
	const { handleOpenDateModal } = useUiActions();
	const [lastView] = useState(initialStateLastView);
	const { events, handleSetActiveEvent, startLoadingEvents } =
		useCalendarActions();

	useEffect(() => {
		startLoadingEvents();
	}, []);

	const eventStyleGetter = () => {
		//PROPS: event, start, end, isSelected

		const style = {
			backgroundColor: '#347CF7',
			borderRadius: '0px',
			opacity: 0.8,
			color: 'white',
		};

		return {
			style,
		};
	};

	const onDobleClick = () => {
		handleOpenDateModal();
	};

	const onSelect = (event) => {
		handleSetActiveEvent(event);
	};

	const onViewChanged = (event) => {
		localStorage.setItem('lastView', event);
	};

	return (
		<>
			<NavBar />
			<Calendar
				culture='es'
				localizer={localizer}
				defaultView={lastView}
				events={events}
				startAccessor='start'
				endAccessor='end'
				style={{ height: 'calc(100vh - 80px)' }}
				messages={getMessagesEs()}
				eventPropGetter={eventStyleGetter}
				components={{
					event: CalendarEvent,
				}}
				onDoubleClickEvent={onDobleClick}
				onSelectEvent={onSelect}
				onView={onViewChanged}
			/>
			<CalendarModal />

			<FabAddNew />

			<FabDelete />
		</>
	);
};
