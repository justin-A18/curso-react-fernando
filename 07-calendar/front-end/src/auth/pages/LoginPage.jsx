import { ToastContainer, toast } from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';

import { Login, Register } from '../components';
import './LoginPage.css';
import { useEffect } from 'react';
import { useAuthActions } from '../../hooks/useAuthActions';

export const LoginPage = () => {
	const { errorMessage } = useAuthActions();

	useEffect(() => {
		if (errorMessage) {
			const err = Object.values(errorMessage);

			if (err.length > 0 && typeof errorMessage !== 'string') {
				for (const iterador of err) {
					toast.error(iterador.msg, {
						theme: 'colored',
					});
				}

				return;
			}

			toast.error(errorMessage, {
				theme: 'colored',
			});
		}
	}, [errorMessage]);

	return (
		<>
			<ToastContainer />
			<div className='container login-container'>
				<div className='row'>
					<Login />
					<Register />
				</div>
			</div>
		</>
	);
};
