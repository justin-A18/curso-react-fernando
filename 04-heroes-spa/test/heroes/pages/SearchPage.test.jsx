import { describe, test, expect, jest, beforeEach } from '@jest/globals';
import { fireEvent, render, screen } from '@testing-library/react';
import { SearchPage } from '../../../src/heroes/pages/SearchPage';
import { MemoryRouter } from 'react-router-dom';

const mockedUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
	...jest.requireActual('react-router-dom'),
	useNavigate: () => mockedUseNavigate,
}));

describe('Pruebas en <SearchPage/>', () => {
	beforeEach(() => jest.clearAllMocks());

	test('Debe de mostrarse correctamente con valores por defecto', () => {
		//CRAER EL SNAPSHOT
		const { container } = render(
			<MemoryRouter>
				<SearchPage />
			</MemoryRouter>
		);

		expect(container).toMatchSnapshot();
	});

	test('Debe de mostrar a Batman y el input con el valor del queryString', () => {
		//CRAER EL SNAPSHOT
		render(
			<MemoryRouter initialEntries={['/search?q=batman']}>
				<SearchPage />
			</MemoryRouter>
		);

		const input = screen.getByRole('textbox');
		expect(input.value).toBe('batman');

		const img = screen.getByRole('img');
		expect(img.src).toBe('http://localhost/heroes/dc-batman.jpg');

		const searchDiv = screen.getByLabelText('search-hero');
		expect(searchDiv.style.display).toBe('none');
	});

	test('Debe de mostrar un error si no se encuentra el hero', () => {
		render(
			<MemoryRouter initialEntries={['/search?q=superhero']}>
				<SearchPage />
			</MemoryRouter>
		);

		const alertDanger = screen.getByLabelText('alert-danger');
		expect(alertDanger.style.display).toBe('');
	});

	test('Debe llamar el navigate a la pantalla nueva', () => {
		render(
			<MemoryRouter initialEntries={[`/search`]}>
				<SearchPage />
			</MemoryRouter>
		);

		const inputSearch = screen.getByRole('textbox');
		fireEvent.change(inputSearch, {
			target: { name: 'searchText', value: 'superman' },
		});

		const formSearch = screen.getByLabelText('form-search');
		fireEvent.submit(formSearch);

		expect(mockedUseNavigate).toHaveBeenCalledWith('?q=superman');
	});
});
