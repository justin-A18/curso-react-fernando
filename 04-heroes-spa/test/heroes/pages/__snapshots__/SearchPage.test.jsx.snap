// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchPage/> Debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form
        aria-label="form-search"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        aria-label="search-hero"
        class="alert alert-primary"
      >
        Search a hero
      </div>
      <div
        aria-label="alert-danger"
        class="alert alert-danger"
        style="display: none;"
      >
        No hero with 
        <b />
      </div>
    </div>
  </div>
</div>
`;
